File: src/main/java/io/github/katrix_/chitchat/chat/ChatListener.java
/**
 * This file is part of ChitChat, licensed under the MIT License (MIT).
 *
 * Copyright (c) 2016 Katrix
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
 * associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute,
 * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
 * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package io.github.katrix_.chitchat.chat;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import org.spongepowered.api.effect.sound.SoundTypes;
import org.spongepowered.api.entity.living.player.Player;
import org.spongepowered.api.event.Listener;
import org.spongepowered.api.event.Order;
import org.spongepowered.api.event.filter.cause.First;
import org.spongepowered.api.event.message.MessageChannelEvent;
import org.spongepowered.api.event.message.MessageEvent.MessageFormatter;
import org.spongepowered.api.event.network.ClientConnectionEvent;
import org.spongepowered.api.service.permission.Subject;
import org.spongepowered.api.service.permission.option.OptionSubject;
import org.spongepowered.api.text.Text;
import org.spongepowered.api.text.TextElement;
import org.spongepowered.api.text.TranslatableText;
import org.spongepowered.api.text.channel.MessageChannel;
import org.spongepowered.api.text.channel.type.CombinedMessageChannel;
import org.spongepowered.api.text.format.TextFormat;
import org.spongepowered.api.text.format.TextStyle;
import org.spongepowered.api.text.serializer.FormattingCodeTextSerializer;
import org.spongepowered.api.text.serializer.TextSerializers;
import org.spongepowered.api.text.transform.SimpleTextTemplateApplier;

import com.google.common.collect.ImmutableMap;

import io.github.katrix_.chitchat.ChitChat;
import io.github.katrix_.chitchat.helper.LogHelper;
import io.github.katrix_.chitchat.io.ConfigSettings;
import io.github.katrix_.chitchat.lib.LibPerm;

public class ChatListener {

	@Listener(order = Order.LATE)
	public void onChat(MessageChannelEvent.Chat event, @First Player player) {
		ConfigSettings cfg = ChitChat.getConfig();
		MessageFormatter formatter = event.getFormatter();
		FormattingCodeTextSerializer serializer = TextSerializers.FORMATTING_CODE;
		Text prefix = cfg.getDefaultPrefix();
		Text suffix = cfg.getDefaultSuffix();
		TextElement name = getNameChat(formatter.getHeader().getAll(), "header").orElse(Text.of(player.getName()));

		Subject subject = player.getContainingCollection().get(player.getIdentifier());
		if(subject instanceof OptionSubject) {
			OptionSubject optionSubject = (OptionSubject)subject;
			Optional<String> option;

			option = optionSubject.getOption("prefix");
			prefix = option.isPresent() ? serializer.deserialize(option.get()) : prefix;

			option = optionSubject.getOption("suffix");
			suffix = option.isPresent() ? serializer.deserialize(option.get()) : suffix;

			option = optionSubject.getOption("nameColor");
			Optional<TextFormat> textFormat = option.flatMap(s -> {
				List<Text> children = serializer.deserialize(s).getChildren();
				if(!children.isEmpty()) {
					return Optional.of(children.get(0).getFormat());
				}
				else return Optional.empty();
			});
			name = textFormat.isPresent() ? Text.of(textFormat.get(), name) : name;
		}

		for(SimpleTextTemplateApplier applier : formatter.getHeader()) {
			if(applier.getParameters().containsKey("header")) {
				applier.setTemplate(cfg.getHeaderTemplate());
				applier.setParameter("header", name);
			}

			applier.setParameter(ConfigSettings.TEMPLATE_PREFIX, prefix);
		}

		if(player.hasPermission(LibPerm.CHAT_COLOR)) {
			for(SimpleTextTemplateApplier applier : formatter.getBody()) {
				if(applier.getParameters().containsKey("body")) {
					applier.setParameter("body", serializer.deserialize(serializer.serialize(Text.of(applier.getParameter("body"))))); //In case the message is already formatted
				}
			}
		}

		SimpleTextTemplateApplier suffixApplier = new SimpleTextTemplateApplier(cfg.getSuffixTemplate());
		formatter.getFooter().add(suffixApplier);

		for(SimpleTextTemplateApplier applier : formatter.getFooter().getAll()) {
			applier.setParameter(ConfigSettings.TEMPLATE_SUFFIX, suffix);
		}

		ChannelChitChat playerChannel = ChitChatPlayers.getOrCreate(player).getChannel();
		event.setChannel(new CombinedMessageChannel(playerChannel, MessageChannel.TO_CONSOLE));

		if(cfg.getChatPling()) {
			String message = event.getFormatter().getBody().format().toPlain();
			Map<Player, UserChitChat> playerMap = ChitChatPlayers.getMap();
			for(Map.Entry<Player, UserChitChat> entry : playerMap.entrySet()) {
				Player player1 = entry.getKey();
				if(entry.getValue().getChannel().equals(playerChannel) && message.contains(player1.getName())) {
					player1.playSound(SoundTypes.ORB_PICKUP, player1.getLocation().getPosition(), 0.5D);
				}
			}
		}
	}

	@SuppressWarnings("Convert2streamapi")
	@Listener
	public void onJoin(ClientConnectionEvent.Join event) {
		ConfigSettings cfg = ChitChat.getConfig();

		List<SimpleTextTemplateApplier> appliers = event.getFormatter().getBody().getAll();
		TextElement name = getNameConnect(appliers, "body").orElse(Text.of(event.getTargetEntity().getName()));

		for(SimpleTextTemplateApplier applier : appliers) {
			if(applier.getParameters().containsKey("body")) {
				applier.setTemplate(ChitChat.getConfig().getJoinTemplate());
				applier.setParameter("body", name);
			}
		}

		Player player = event.getTargetEntity();
		ChannelChitChat channel = ChitChat.getStorage().getChannelForUser(player);
		ChitChatPlayers.getOrCreate(player).setChannel(channel);
		player.sendMessage(cfg.getChattingJoinTemplate(), ImmutableMap.of(ConfigSettings.TEMPLATE_CHANNEL, Text.of(channel.getName())));
	}

	@SuppressWarnings("Convert2streamapi")
	@Listener
	public void onQuit(ClientConnectionEvent.Disconnect event) {
		List<SimpleTextTemplateApplier> appliers = event.getFormatter().getBody().getAll();
		TextElement name = getNameConnect(appliers, "body").orElse(Text.of(event.getTargetEntity().getName()));

		for(SimpleTextTemplateApplier applier : appliers) {
			if(applier.getParameters().containsKey("body")) {
				applier.setTemplate(ChitChat.getConfig().getDisconnectTemplate());
				applier.setParameter("body", name);
			}
		}
	}

	private Optional<TextElement> getNameChat(List<SimpleTextTemplateApplier> applierList, String parameterName) {
		for(SimpleTextTemplateApplier applier : applierList) {
			if(applier.getParameters().containsKey(parameterName)) {
				TextElement textElement = applier.getParameter(parameterName);
				return Optional.of(textElement);
			}
		}

		return Optional.empty();
	}

	private Optional<Text> getNameConnect(List<SimpleTextTemplateApplier> applierList, String parameterName) {
		for(SimpleTextTemplateApplier applier : applierList) {
			if(applier.getParameters().containsKey(parameterName)) {
				TextElement textElement = applier.getParameter(parameterName);
				if(textElement instanceof TranslatableText) {
					TranslatableText translatable = (TranslatableText)textElement;
					return translatable.getArguments().stream()
							.filter(o -> o instanceof Text)
							.map(o -> (Text)o)
							.findFirst();
				}
			}
		}

		return Optional.empty();
	}
}

File: chitChat/shared/src/main/scala/net/katsstuff/chitchat/chat/ChatListener.scala
package net.katsstuff.chitchat.chat

import scala.collection.JavaConverters._

import org.spongepowered.api.entity.living.player.Player
import org.spongepowered.api.event.entity.living.humanoid.player.TargetPlayerEvent
import org.spongepowered.api.event.message.{MessageChannelEvent, MessageEvent}
import org.spongepowered.api.event.network.ClientConnectionEvent
import org.spongepowered.api.event.{Listener, Order}
import org.spongepowered.api.text.channel.MessageChannel
import org.spongepowered.api.text.channel.`type`.CombinedMessageChannel
import org.spongepowered.api.text.format.TextColors._
import org.spongepowered.api.text.serializer.TextSerializers
import org.spongepowered.api.text.transform.SimpleTextTemplateApplier
import org.spongepowered.api.text.{transform, Text, TextElement, TranslatableText}

import net.katsstuff.chitchat.ChatConfig.{ChatFormatter, ChatOption, FormatterOptions}
import net.katsstuff.chitchat.helper.TextHelper
import net.katsstuff.chitchat.lib.LibPerm
import net.katsstuff.chitchat.{ChatConfig, ChitChatPlugin}
import net.katsstuff.katlib.helper.Implicits._

class ChatListener(implicit plugin: ChitChatPlugin, handler: ChannelHandler) {

  private val oldSerializer  = TextSerializers.FORMATTING_CODE
  private val jsonSerializer = TextSerializers.JSON

  private var cfg: ChatConfig.RootConfig = _

  /**
    * Reloads the config used by this listener
    */
  def reloadConfig(config: ChatConfig.RootConfig): Unit = cfg = config

  //Search for a option to use, trying both json and formatting codes
  def getTextOption(player: Player, variant: String): Option[Text] =
    player
      .getOption(s"json$variant")
      .toOption
      .map(jsonSerializer.deserialize)
      .orElse {
        player
          .getOption(variant)
          .toOption
          .map(oldSerializer.deserialize)
      }

  @Listener(order = Order.LAST)
  def onMessageChannel(event: MessageChannelEvent.Chat): Unit = {
    event.getCause.first(classOf[Player]).toOption match {
      case Some(player) =>
        val formatter = event.getFormatter
        val name      = findNameChat(formatter.getHeader.getAll.asScala, "header").getOrElse(t"${player.getName}")

        val stringFormat  = getTextOption(player, "namestyle")
        val textFormat    = stringFormat.flatMap(text => TextHelper.getFormatAtEnd(text))
        val formattedName = textFormat.fold(name)(format => t"$format$name")

        //Colors the body from formatting codes. This might destroy other data
        if (player.hasPermission(LibPerm.ChatColor)) {
          for (applier <- formatter.getBody.asScala if applier.getParameters.containsKey("body")) {
            applier.getParameter("body") match {
              case body: Text if body.toPlain.contains('&') =>
                applier.setParameter("body", oldSerializer.deserialize(oldSerializer.serialize(body)))
              case _ =>
            }
          }
        }

        formatEvent(event, player, cfg.formatting.chat, formattedName)

        event.getChannel.toOption.foreach { existing =>
          //Apply chat channel
          val playerChannel = handler.getChannelForReceiver(player)
          event.setChannel(new IntersectionMessageChannel(Set(existing, playerChannel.messageChannel)))

          //Pling mention
          if (cfg.mentionPling) {
            val message = event.getFormatter.getBody.format.toPlain
            playerChannel.members
              .flatMap(_.get)
              .collect { case player: Player if message.contains(player.getName) => player }
              .foreach(p => p.playSound(plugin.versionHelper.levelUpSound, p.getLocation.getPosition, 0.5D))
          }
        }
      case None =>
    }

    if (event.getCause.contains(SendToConsole)) {
      event.getChannel.toOption.foreach(c => event.setChannel(new CombinedMessageChannel(c, MessageChannel.TO_CONSOLE)))
    }
  }

  @Listener
  def onJoin(event: ClientConnectionEvent.Join): Unit = {
    val player = event.getTargetEntity
    handler.loginPlayer(player)
    formatConnectEvent(event, cfg.formatting.join)
    player.sendMessage(t"${YELLOW}You are chatting in ${handler.getChannelForReceiver(player).name}")
  }

  @Listener
  def onDisconnect(event: ClientConnectionEvent.Disconnect): Unit =
    formatConnectEvent(event, cfg.formatting.disconnect)

  def formatConnectEvent(
      event: ClientConnectionEvent with MessageEvent with TargetPlayerEvent,
      chatFormatter: ChatFormatter
  ): Unit = {
    val player = event.getTargetEntity
    val name   = findNameConnect(event.getFormatter.getBody.asScala.toSeq, "body").getOrElse(t"${player.getName}")

    val stringFormat  = getTextOption(player, "namestyle")
    val textFormat    = stringFormat.flatMap(text => TextHelper.getFormatAtEnd(text))
    val formattedName = textFormat.fold(name)(format => t"$format$name")

    formatEvent(event, player, chatFormatter, formattedName)
  }

  def formatEvent(
      event: MessageEvent,
      player: Player,
      chatFormatter: ChatFormatter,
      formattedName: TextElement
  ): Unit = {
    val formatter = event.getFormatter
    val sections = Map(
      formatter.getHeader -> chatFormatter.header,
      formatter.getBody   -> chatFormatter.body,
      formatter.getFooter -> chatFormatter.footer
    )

    for {
      (section, FormatterOptions(_, newTemplates, _)) <- sections
      newTemplate                                     <- newTemplates
    } {
      section.add(new transform.SimpleTextTemplateApplier(newTemplate))
    }

    for {
      (section, FormatterOptions(overrideTemplates, _, parameters)) <- sections
      applier                                                       <- section.asScala
    } {
      overrideTemplates.foreach {
        case (key, template) =>
          if (applier.getParameters.containsKey(key)) {
            applier.setTemplate(template)
          }
      }

      parameters.foreach {
        case (confName, ChatOption(optionName, isText, default)) =>
          if (optionName == "namestyle") {
            applier.setParameter("header", formattedName)
          }
          val content =
            if (isText) getTextOption(player, optionName).getOrElse(default)
            else player.getOption(optionName).toOption.fold(default)(Text.of)
          applier.setParameter(confName, content)
      }
    }
  }

  private def findNameChat(applierList: Seq[SimpleTextTemplateApplier], parameterName: String): Option[TextElement] =
    findNameCommon(applierList, parameterName).headOption

  private def findNameConnect(applierList: Seq[SimpleTextTemplateApplier], parameterName: String): Option[Text] =
    findNameCommon(applierList, parameterName).collectFirst {
      case translatable: TranslatableText => translatable.getArguments.asScala.collectFirst { case t: Text => t }
    }.flatten

  private def findNameCommon(applierList: Seq[SimpleTextTemplateApplier], parameterName: String): Seq[TextElement] =
    applierList
      .withFilter(_.getParameters.containsKey(parameterName))
      .map(_.getParameter(parameterName))
}

object SendToConsole
